[Flang][OpenMP] Add some semantic checks for Linear clause

Thirumalai-Shaktivel opened the pull request:
This PR adds all the missing semantics for the Linear clause based on the OpenMP 5.2 restrictions. The restriction details are mentioned below.

OpenMP 5.2:
5.4.6 linear Clause restrictions
- A linear-modifier may be specified as ref or uval only on a declare simd directive.
- If linear-modifier is not ref, all list items must be of type integer.
- If linear-modifier is ref or uval, all list items must be dummy arguments without the VALUE attribute.
- List items must not be Cray pointers or variables that have the POINTER attribute. Cray pointer support has been deprecated.
- If linear-modifier is ref, list items must be polymorphic variables, assumed-shape arrays, or variables with the ALLOCATABLE attribute. 
- A common block name must not appear in a linear clause.
- The list-item cannot appear more than once

4.4.4 ordered Clause restriction
- If n is explicitly specified, a linear clause must not be specified on the same directive.

5.11 aligned Clause restriction
- Each list item must have C_PTR or Cray pointer type or have the POINTER or ALLOCATABLE attribute. Cray pointer support has been deprecated.

Thirumalai-Shaktivel committed:
[LLVM-Flang][OpenMP] Add some semantic checks for Linear clause based on the OpenMP 5.2 restrictions

Thirumalai-Shaktivel committed:
[LLVM-Flang][OpenMP] Skip Linear clause semantic check if ordered clause with (n) is specified

Thirumalai-Shaktivel committed:
[Test] Add a test and update a check

kparzysz commented on flang/test/Semantics/OpenMP/linear-cluase01.f90:
Typo: cluase -> clause.

Thirumalai-Shaktivel replied:
Done.

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
 void OmpStructureChecker::Enter(const parser::OmpClause::Linear &x) {
   CheckAllowedClause(llvm::omp::Clause::OMPC_linear);
 
+  parser::CharBlock source = GetContext().clauseSource;
```
Use braced initialized in frontend code.

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
     }
   }
+
+  auto checkForValidLinearClaue = [&](const parser::Name &name, bool is_ref) {
```
```suggestion
  auto checkForValidLinearClause = [&](const parser::Name &name, bool is_ref) {
```

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+
+  auto checkForValidLinearClaue = [&](const parser::Name &name, bool is_ref) {
+    parser::CharBlock source = GetContext().clauseSource;
+    std::string list_item_name = name.ToString();
```
```suggestion
    std::string listItemName = name.ToString();
```

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
   }
+
+  auto checkForValidLinearClaue = [&](const parser::Name &name, bool is_ref) {
+    parser::CharBlock source = GetContext().clauseSource;
```
Nit: Use braced initialization.

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+                checkForValidLinearClaue(name,
+                    (withModifier.modifier.v ==
+                        parser::OmpLinearModifier::Type::Ref));
+                std::string list_item_name = name.ToString();
```
```suggestion
                std::string listItemName = name.ToString();
```

kiranchandramohan commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -303,6 +303,30 @@ void OmpStructureChecker::CheckMultListItems() {
+
+  // Linear clause
+  auto linearClauses{FindClauses(llvm::omp::Clause::OMPC_linear)};
+  for (auto itr = linearClauses.first; itr != linearClauses.second; ++itr) {
```
```suggestion
  for (auto itr{linearClauses.first}; itr != linearClauses.second; ++itr) {
```

kiranchandramohan reviewed (commented):
A few nit comments.

Could you expand the summary to list the checks implemented in the patch with a reference to the standard and section?

Thirumalai-Shaktivel commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -2975,16 +3000,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+          FindClause(GetContext(), llvm::omp::Clause::OMPC_ordered)}) {
+    const auto &orderedClause{std::get<parser::OmpClause::Ordered>(clause->u)};
+    if (orderedClause.v) {
+      return;
```
Flang-new already throws an error, `Clause LINEAR is not allowed if clause ORDERED appears on the DO directive`. So, I skipped here.

Thirumalai-Shaktivel commented:
I addressed all the comments, @kiranchandramohan.

Thanks for the reviews.

Thirumalai-Shaktivel committed:
[Test] Fix a test to be valid

OpenMP 5.2 restriction: Each list item must have C_PTR or Cray pointer type or have the POINTER or ALLOCATABLE attribute.

NimishMishra commented on flang/test/Semantics/OpenMP/declarative-directive.f90:
```diff
@@ -23,6 +23,7 @@ end subroutine requires_2
 
 subroutine declare_simd_1(a, b)
   real(8), intent(inout) :: a, b
+  !ERROR: 'a' in ALIGNED clause must be of type C_PTR, POINTER or ALLOCATABLE
```
Do we also deal with ALIGNED clause in this PR? A similar issue is causing the Linux CI failure:

```
flang/test/Examples/omp-declarative-directive.f90:11:38: error: 'a' in ALIGNED clause must be of type C_PTR, POINTER or ALLOCATABLE
--
  | !$omp declare simd(declare_simd_1) aligned(a)
```

Thirumalai-Shaktivel replied:
Yes, this line: https://github.com/llvm/llvm-project/pull/111354/files#diff-be225bc038e95a229024c9f227284112f98fe31762bc0bb80605de425aadd084R2287 is responsible for the change.

Thirumalai-Shaktivel replied:
I fixed it in the recent commit.

NimishMishra reviewed (commented):
One small comment on ALIGNED clause test added in this PR.

Ultra nit: "If n is explicitly specified, a linear clause must not be specified on the same directive" in description -> "If an ordered clause is explicitly specified......"

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -323,6 +323,30 @@ void OmpStructureChecker::CheckMultListItems() {
   }
+
+  // Linear clause
+  auto linearClauses{FindClauses(llvm::omp::Clause::OMPC_linear)};
+  for (auto itr{linearClauses.first}; itr != linearClauses.second; ++itr) {
+    const auto &linearClause{
+        std::get<parser::OmpClause::Linear>(itr->second->u)};
```
`FindClauses` now returns an iterator range, please use range-for here.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -323,6 +323,30 @@ void OmpStructureChecker::CheckMultListItems() {
+        std::get<parser::OmpClause::Linear>(itr->second->u)};
+    std::list<parser::Name> nameList;
+    common::visit(
+        common::visitors{
+            [&](const parser::OmpLinearClause::WithoutModifier
+                    &withoutModifier) {
+              for (const auto &name : withoutModifier.names) {
+                nameList.push_back(name);
+              }
+            },
+            [&](const parser::OmpLinearClause::WithModifier &withModifier) {
+              for (const auto &name : withModifier.names) {
+                nameList.push_back(name);
+              }
+            },
+        },
```
You can only have a single visitor, plus you can simplify the copy:
```
[&](auto &&s) {
  std::copy(s.names.begin(), s.names.end(), std::back_inserter(nameList));
}
```

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3040,16 +3064,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+          },
+          [&](const parser::OmpLinearClause::WithModifier &withModifier) {
+            for (const auto &name : withModifier.names) {
+              if (name.symbol) {
```
You can extract this into a lamda to reduce indentation.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3040,16 +3064,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+                        parser::OmpLinearModifier::Type::Val &&
+                    IsDummy(*name.symbol) && IsValue(*name.symbol)) {
+                  context_.Say(source,
+                      "The list item `%s` specified with the linear-modifier "
+                      "`REF` or `UVAL` must be a dummy argument without "
+                      "`VALUE` attribute"_err_en_US,
```
Please concatenate messages into a single string.  This is to make grepping for messages easier.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3040,16 +3064,101 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+                        IsAssumedShape(*name.symbol) ||
+                        IsPolymorphic(*name.symbol))) {
+                  context_.Say(source,
+                      "The list item `%s` specified with the linear-modifier "
+                      "`REF` must be polymorphic variable, assumed-shape "
+                      "array, "
+                      "or a variable with the `ALLOCATABLE` attribute"_err_en_US,
```
Same here.

kparzysz commented:
I will change the modifier representation in LINEAR in a week or so.  Please finish this soon, otherwise my code will create conflicts for you.

Thirumalai-Shaktivel commented:
Thanks for the review, @kparzysz!

I addressed all the comments. If anything else to be done, please let me know.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3542,16 +3557,98 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+  common::visit(
+      common::visitors{
+          [&](const parser::OmpLinearClause::WithoutModifier &withoutModifier) {
+            for (const auto &name : withoutModifier.names)
+              if (name.symbol)
+                checkForValidLinearClause_01(name, false);
```
Please add {} for nested statements.

Thirumalai-Shaktivel replied:
Do you mean the following?
```suggestion
            for (const auto &name : withoutModifier.names) {
              if (name.symbol) {
                checkForValidLinearClause_01(name, false);
              }
            }
```

kparzysz replied:
Yes

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -383,6 +383,21 @@ void OmpStructureChecker::CheckMultListItems() {
+    const auto &linearClause{std::get<parser::OmpClause::Linear>(clause->u)};
+    std::list<parser::Name> nameList;
+    common::visit(
+        common::visitors{
+            [&](const auto &u) {
+              std::copy(
+                  u.names.begin(), u.names.end(), std::back_inserter(nameList));
+            },
+        },
```
You only have one visitor, so you don't need `common::visitors`.  Do `common::visit([&]...)`.

Thirumalai-Shaktivel replied:
Yea, sure.

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3542,16 +3557,98 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
     }
   }
+
+  auto checkForValidLinearClause_01 = [&](const parser::Name &name,
+                                          bool is_ref) {
```
Please concatenate all error message strings in this function.

Thirumalai-Shaktivel replied:
I'm sorry, but I'm not getting this.

Concatenate in the sense, for the following:
```
context_.Say(source,
          "The list item `%s` specified with other than linear-modifier `REF`"
          " must be of type INTEGER"_err_en_US,
          listItemName);
```
should I do?
```
context_.Say(source,
          "The list item `%s` specified with other than linear-modifier `REF` must be of type INTEGER"_err_en_US,
          listItemName);
```

kparzysz replied:
Yes

kparzysz commented on flang/lib/Semantics/check-omp-structure.cpp:
```diff
@@ -3542,16 +3557,98 @@ void OmpStructureChecker::Enter(const parser::OmpClause::If &x) {
+
+  auto checkForValidLinearClause_02 =
+      [&](const parser::Name &name,
+          const parser::OmpLinearModifier::Value &modifierValue) {
```
Please concatenate all error message strings in this function.

kparzysz reviewed (commented):
Almost ready, a few more comments.

kparzysz commented:
Please don't use force-pushes.  That messes up previous comments.

Thirumalai-Shaktivel commented:
Sorry about that.

I rebased the commits and I had to force-push the changes.
Next time I will make sure not to rebase it.

Thirumalai-Shaktivel commented:
Done.

Thirumalai-Shaktivel commented:
Thank you!
